Full testing

jamadden opened the pull request:
Fixes #1 

Plus various product modernizations as detailed in the commits.

jamadden committed:
Modernize project structure

- Modern .trovis.yml and tox.ini
- Modern python versions
- Rename .txt -> .rst

jamadden committed:
Fix installing zc.recipe.egg on travis.

Also cleanup our rst and make sure the default matches this.

jamadden committed:
Add tests. The layout file is currently wrong and fails to build.

jamadden committed:
Fix layout.html to be valid syntax.

jamadden committed:
Get tox working

jamadden committed:
We need a no cover here now that layout.html is fixed: we don't expect a bad code

mgedmin commented on README.rst:
```diff
@@ -0,0 +1 @@
+See src/z3c/recipe/sphinxdoc/index.txt for more information.
```
`.txt` or `.rst`?

jamadden replied:
Whoops, missed one.

mgedmin commented on src/z3c/recipe/sphinxdoc/__init__.py:
```diff
@@ -97,8 +98,9 @@ def install(self):
-                f = self.openfile(self.options['default.css'])
-                open(join(staticDir,'default.css'), 'w').write(f.read())
+                with self.openfile(self.options['default.css']) as f:
+                    with open(join(staticDir,'default.css'), 'w') as w:
```
Space after comma?

mgedmin commented on src/z3c/recipe/sphinxdoc/__init__.py:
```diff
@@ -111,8 +113,9 @@ def install(self):
-                f = self.openfile(self.options['layout.html'])
-                open(join(templatesDir,'layout.html'), 'w').write(f.read())
+                with self.openfile(self.options['layout.html']) as f:
+                    with open(join(templatesDir,'layout.html'), 'w') as w:
```
Space after comma?

mgedmin commented on src/z3c/recipe/sphinxdoc/__init__.py:
```diff
@@ -168,9 +171,11 @@ def install(self):
 
 
-def main(projects):
+def main(projects, argv=None, exit_on_error=False):
```
Shouldn't exit_on_error default to True?

jamadden replied:
I don't *think* so? This recipe is used for generating the docs for many projects at once. I don't think you want it to bail after generating doc number 3 of 50 just because doc number 3 happened to have an error.

It didn't behave that way in the past, AFAICS.

jamadden replied:
I don't have an issue for this yet, but our fork of this project goes a step further and needs to handle [bad eggs at script creation time](https://github.com/NextThought/z3c.recipe.sphinxdoc/commit/a09782462f00993817d4f4dc1e35826b275d9177). 

I think this recipe is a "best effort" recipe, not a "fail fast" or "bondage-and-discipline-warnings-are-errors" recipe.

mgedmin replied:
I didn't check, but I thought `sphinx.main()` that was there before would exit with an non-zero exit code on error?

As far as I can see, this `main()` is invoked as an entry point, from a new script generated in `bin/`, and I think this script should exit with non-zero if documentation generation fails.

mgedmin replied:
I was wrong: `sphinx.main()` doesn't call `sys.exit()`, it just returns the result of `sphinx.build_main()`.

Never mind!

jamadden replied:
There was actually a brief period of time where `sphinx.main` *did* exit. That introduced a bug in this package (#2). But later sphinx removed it again.

jamadden replied:
> and I think this script should exit with non-zero if documentation generation fails.

I wouldn't argue with that, *if* it tried to generate all the docs before exiting. So something like:
```python
code = 0
for project in projects:
   code |= sphinx.build()
sys.exit(code)
```

(Except I find it obnoxious when functions call `sys.exit` because it makes them annoying to unittest. I tend to reserve that for the `if __name__ == '__main__':` block. Which doesn't quite work here.)

mgedmin commented on src/z3c/recipe/sphinxdoc/tests/test_sphinxdoc.py:
```diff
@@ -0,0 +1,127 @@
+        docs.install()
+
+
+        html_path = os.path.join(self.templates_path, 'layout.html')
```
I tried to avoid commenting on the number of blank lines in these test methods, but couldn't help myself.

There are a lot of blank lines in these test methods!

jamadden replied:
Oh come on, I have a grand total of three "excess" blank lines, only one of which is *in* a test method 😛 

I think they arise as a result of refactoring as the tests evolve. I guess I just don't see them...

mgedmin replied:
My text editor shows flake8 warnings inline, so I get ugly red markers when there's more than one blank line between methods or less than two between classes/global functions.

I could say this trained me to pay attention to blank lines, but that would be a lie - I was always pedantic about them. :-)

jamadden replied:
My editor does that too (only I use pylint). But I have all the line spacing warnings turned off 😄 I'm all for pedantic, really, but "two lines here", "one line here", etc, really just got on my nerves when I started (coming from a language where there wasn't a style guide for that).

Same thing for the "there must be two spaces between the end of a line and the inline comment" warning. Really? Who actually does that?

mgedmin commented on tox.ini:
```diff
@@ -0,0 +1,24 @@
+    coverage run -m zope.testrunner --test-path=src []
+deps =
+    --no-binary
+     zc.recipe.egg
```
I think it would be clearer if you did it on one line, with `--no-binary=zc.recipe.egg`.

It would be even less hacky if you used `install_command = pip install --no-binary=zc.recipe.egg {opts} {packages}`

jamadden replied:
Hmm, I didn't realize `--no-binary` took an `=`. The help description (`--no-binary <format_control>`) doesn't imply that.

I'm not sure I like the "repetition" in the latter, but I'll play with this.

jamadden replied:
TIL about `install_command`. You were right, that does look better. Thanks!

mgedmin replied:
You're one of the [lucky 10000](https://xkcd.com/1053/) today!

The GNU getopt_long(3), which introduced long options, allows both `--option value` and `--option=value`.  Python's optparse/argparse also support both spellings (and this is even [documented](https://docs.python.org/3/library/argparse.html#option-value-syntax) deep in the reference manual that nobody reads for fun).

I've been used to this for so long that I'm always surprised when I discover a program (*cough* gnome-session-inhibit *cough*) that doesn't accept --foo=bar.

mgedmin reviewed (approved):
LGTM with minor comments

jamadden committed:
Cleanups from feedback

- whitespace in __init__.py
-- Go ahead and refactor duplicate code into own method
- whitespace is test_sphinxdoc.py
- cleaner way to specify no-binary for tox.ini
- correct path in README.rst

jamadden commented:
Thank you!
